### PR TITLE
Change API response schema to be a valid GeoJSON 

### LINKED ISSUE #46: CLI/API output/response should have the same schema
CLI and API should both output a JSON (when input is a feature of a dataset) or GeoJSON (if input is bpolys) with a shared schemata.

### LINKED ISSUE #88: Check API support for arbitrary GeoJSON input such as FeatureCollection
`api.py.` should use `loads_bpolys()` function of `helpers.py` module as well. Currently only used by `cli.py`. This will make `load_bpolys` function of `api.py` obsolete.

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Corresponding issue
Closes #46 
Closes #88 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Check if #88 will be fixed with this PR

### CODE REVIEW

File: workers/ohsome_quality_analyst/oqt.py [hunk 1]
```diff
@@ -3,20 +3,115 @@
...
 import ohsome_quality_analyst.geodatabase.client as db_client
 from ohsome_quality_analyst.base.indicator import BaseIndicator
 from ohsome_quality_analyst.base.report import BaseReport
-from ohsome_quality_analyst.utils.definitions import INDICATOR_LAYER
-from ohsome_quality_analyst.utils.helper import name_to_class
+from ohsome_quality_analyst.utils.definitions import GEOM_SIZE_LIMIT, INDICATOR_LAYER
+from ohsome_quality_analyst.utils.helper import loads_geojson, name_to_class
+
+
+async def create_indicator_as_geojson(
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/oqt.py:18]: Currently this function is only tested indirectly but completly through API and CLI tests.

Not sure if we need additional dedictated tests. If so I would like to change the test structure in a way to make it easy to cover all possible input parameters for "compute indicator/report" functions in `oqt.py`, CLI and API. They all have very similar input parameter.

File: workers/ohsome_quality_analyst/oqt.py [hunk 2]
```diff
@@ -3,20 +3,115 @@
...
+) -> Union[Feature, FeatureCollection]:
+    """Create an indicator or multiple indicators as GeoJSON object.
+
+    Depending on the input a single indicator as GeoJSON Feature will be returned
+    or multiple indicators as GeoJSON FeatureCollection will be returned.
+    """
+    if bpolys is not None:
+        features = []
+        for i, feature in enumerate(loads_geojson(bpolys)):
+            logging.info("Input feature index:\t" + str(i))
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:36]: Why `\t`?

File: workers/ohsome_quality_analyst/oqt.py [hunk 3]
```diff
@@ -3,20 +3,115 @@
...
+) -> Union[Feature, FeatureCollection]:
+    """Create a report multiple reports as GeoJSON object.
+
+    Depending on the input a single report as GeoJSON Feature will be returned
+    or multiple reports as GeoJSON FeatureCollection will be returned.
+    """
+    if bpolys is not None:
+        features = []
+        for i, feature in enumerate(loads_geojson(bpolys)):
+            logging.info("Input feature index:\t" + str(i))
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:84]: Why `\t`?

File: workers/ohsome_quality_analyst/oqt.py [hunk 4]
```diff
@@ -60,19 +155,19 @@ async def from_database(dataset, feature_id) -> bool:
         except UndefinedTableError:
             return False
 
-    indicator_class = name_to_class(class_type="indicator", name=indicator_name)
+    indicator_class = name_to_class(class_type="indicator", name=name)
     logging.info("Creating indicator ...")
-    logging.info("Indicator name:\t" + indicator_name)
+    logging.info("Indicator name:\t" + name)
     logging.info("Layer name:\t" + layer_name)
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:161]:
```suggestion
    logging.info("Indicator name: " + name)
    logging.info("Layer name:     " + layer_name)
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 5]
```diff
@@ -3,20 +3,115 @@
...
+
+async def create_indicator_as_geojson(
+    name: str,
+    layer_name: str,
+    bpolys: Optional[str] = None,
+    dataset: Optional[str] = None,
+    feature_id: Optional[str] = None,
+    fid_field: Optional[str] = None,
+    force: bool = False,
+    size_restriction=False,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:26]: to be consistent
```suggestion
    size_restriction: bool = False,
```

File: workers/ohsome_quality_analyst/oqt.py [hunk 6]
```diff
@@ -3,20 +3,115 @@
...
+
+
+async def create_report_as_geojson(
+    name: str,
+    bpolys: Optional[str] = None,
+    dataset: Optional[str] = None,
+    feature_id: Optional[str] = None,
+    fid_field: Optional[str] = None,
+    force: bool = False,
+    size_restriction=False,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:74]: to be consistent:
```suggestion
    size_restriction: bool = False,
```

File: workers/ohsome_quality_analyst/oqt.py (hunk 1 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/oqt.py:18]: Ideally the function would be tested explicitly in unittests, but you want to change something fundamental first, you can change that in another PR.

File: workers/ohsome_quality_analyst/oqt.py [hunk 7]
```diff
@@ -117,37 +246,9 @@ async def create_all_indicators(force: bool = False) -> None:
...
-            feature=report.feature,
-            dataset=report.dataset,
-            feature_id=report.feature_id,
-            fid_field=fid_field,
-            force=force,
+async def check_area_size(geom: Union[Polygon, MultiPolygon]):
+    if await db_client.get_area_of_bpolys(geom) > GEOM_SIZE_LIMIT:
+        raise ValueError(
+            "Input GeoJSON Object is too big. "
+            + "The area should be less than {0} sqkm.".format(GEOM_SIZE_LIMIT)
```
Comment (joker234) [inline workers/ohsome_quality_analyst/oqt.py:253]:
```suggestion
            "The area should be less than {0} sqkm.".format(GEOM_SIZE_LIMIT)
```

Comment (joker234) [reply workers/ohsome_quality_analyst/oqt.py:253]: Another suggestion:
```suggestion
            "The area should be less than {0} km².".format(GEOM_SIZE_LIMIT)
```

File: workers/tests/integrationtests/test_api_report.py [hunk 8]
```diff
@@ -5,159 +5,120 @@
...
+        report.set_indicator_layer()
+        self.number_of_indicators = len(report.indicator_layer)
 
+    def run_tests(self, response) -> None:
         self.assertEqual(response.status_code, 200)
-
-        report = response.json()
-        self.schema.validate(report)  # Print information if validation fails
-        self.assertTrue(self.schema.is_valid(report))
+        response = response.json()
```
Comment (joker234) [inline workers/tests/integrationtests/test_api_report.py:40]: That's not very pretty to overwrite the given parameter. Can be sometimes problematic even though in this case not problematic. I would use `json_response` or something else that comes handy to you.

File: workers/tests/integrationtests/test_api_report_geojson_io.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        self.feature_schema = get_feature_schema()
+        self.featurecollection_schema = get_featurecollection_schema()
+
+        report = SimpleReport()
+        report.set_indicator_layer()
+        self.number_of_indicators = len(report.indicator_layer)
+
+    def run_tests(self, response, featurecollection=False) -> None:
+        self.assertEqual(response.status_code, 200)
+        response = response.json()
```
Comment (joker234) [inline workers/tests/integrationtests/test_api_report_geojson_io.py:51]: see above

**Review by joker234 — CHANGES_REQUESTED** [review]
Very good work, especially on the refactoring of the tests. Apart from the few comments I made, I think it would be good to redo all the vcr_cassettes. Some have to be deleted since you renamed tests and the User Agent changed for all requests (see #62).

File: CHANGELOG.md [hunk 10]
```diff
@@ -20,6 +30,7 @@
 [#110]: https://github.com/GIScience/ohsome-quality-analyst/pull/110
 [#112]: https://github.com/GIScience/ohsome-quality-analyst/pull/112
 [#96]: https://github.com/GIScience/ohsome-quality-analyst/pull/96
+[#100]: https://github.com/GIScience/ohsome-quality-analyst/pull/100
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md:33]: Changed the ordering.

File: CHANGELOG.md [hunk 11]
```diff
@@ -6,6 +6,16 @@
...
 
+### Breaking Changes
+
+- Change API response schema ([#100])
+    - Response is a valid GeoJSON
+    - Response equates file output of CLI
+
+### New Features
+
+- FeatureCollection with multiple Features as input to the API ([#100])
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md:17]: Thanks for the verb!

File: CHANGELOG.md [hunk 12]
```diff
@@ -6,6 +6,16 @@
...
+
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md:18]: Alright. I rearranged the sections.

File: CHANGELOG.md [hunk 13]
```diff
@@ -6,6 +6,16 @@
 
 - Fix error raising while serializing JSON due NaN ([#112])
 
+### Breaking Changes
+
+- Change API response schema ([#100])
+    - Response is a valid GeoJSON
+    - Response equates file output of CLI
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply CHANGELOG.md:13]: incorporated

File: workers/ohsome_quality_analyst/oqt.py (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:36]: Exchanged tab for space.

File: workers/ohsome_quality_analyst/oqt.py (hunk 3 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:84]: Exchanged tab for space.

File: workers/ohsome_quality_analyst/oqt.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:161]: Exchanged tab for space.

File: workers/ohsome_quality_analyst/oqt.py (hunk 5 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:26]: Added type hint.

File: workers/ohsome_quality_analyst/oqt.py (hunk 6 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:74]: Added type hint.

File: workers/ohsome_quality_analyst/oqt.py (hunk 7 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:253]: I changed the formatting to you second suggestion.

File: workers/tests/integrationtests/test_api_report.py (hunk 8 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_api_report.py:40]: Yes, true.. I renamed it to response_json.

File: workers/tests/integrationtests/test_api_report_geojson_io.py (hunk 9 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_api_report_geojson_io.py:51]: Renamed it to response_json.

**Comment by matthiasschaub** [comment]
> Very good work, especially on the refactoring of the tests. Apart from the few comments I made, I think it would be good to redo all the vcr_cassettes. Some have to be deleted since you renamed tests and the User Agent changed for all requests (see #62).

Thanks! Good idea. I recreated the VCR cassettes and integrated you suggestions. 
Could you take another look? I left the small commit messages as they are for the review, but will rebase after you viewed the changes again.

File: workers/ohsome_quality_analyst/oqt.py (hunk 1 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/oqt.py:18]: I added to issues for this case:
https://github.com/GIScience/ohsome-quality-analyst/issues/114
https://github.com/GIScience/ohsome-quality-analyst/issues/115

**Review by joker234 — APPROVED** [review]

**Review by joker234 — APPROVED** [review]

**Pull request merged**